ArchiSteamFarm/Statistics.cs

```diff
@@ -457,22 +463,64 @@ internal sealed class Statistics : IDisposable {
 						do {
 							match = false;
 
-							foreach ((ulong ourItem, uint ourAmount) in ourFullSet.Where(item => item.Value > 1).OrderByDescending(item => item.Value)) {
-								if (!ourTradableSet.TryGetValue(ourItem, out uint tradableAmount) || (tradableAmount == 0)) {
+							foreach ((ulong ourItem, uint ourFullAmount) in ourFullSet.Where(item => item.Value > 1).OrderByDescending(item => item.Value)) {
+								if (!ourTradableSet.TryGetValue(ourItem, out uint ourTradableAmount) || (ourTradableAmount == 0)) {
 									continue;
 								}
 
-								foreach ((ulong theirItem, _) in theirItems.OrderBy(item => ourFullSet.TryGetValue(item.Key, out uint ourAmountOfTheirItem) ? ourAmountOfTheirItem : 0)) {
-									if (ourFullSet.TryGetValue(theirItem, out uint ourAmountOfTheirItem) && (ourAmount <= ourAmountOfTheirItem + 1)) {
+								foreach ((ulong theirItem, uint theirFullAmount) in theirFullItems.OrderBy(item => ourFullSet.TryGetValue(item.Key, out uint ourAmountOfTheirItem) ? ourAmountOfTheirItem : 0)) {
+									if (!theirTradableItems.TryGetValue(theirItem, out uint theirTradableAmount) || (theirTradableAmount == 0)) {
+										continue;
+									}
+
+									if (ourFullSet.TryGetValue(theirItem, out uint ourAmountOfTheirItem) && (ourFullAmount <= ourAmountOfTheirItem + 1)) {
 										continue;
 									}
 
+									if (theirFullItems.TryGetValue(ourItem, out uint theirAmountOfOurItem) && (theirFullAmount <= theirAmountOfOurItem + 1)) {
+										continue;
+									}
+
+									if (!listedUser.MatchEverything) {
+										// We have a potential match, let's check fairness for them
+										fairClassIDsToGive[ourItem] = fairClassIDsToGive.TryGetValue(ourItem, out uint fairGivenAmount) ? fairGivenAmount + 1 : 1;
+										fairClassIDsToReceive[theirItem] = fairClassIDsToReceive.TryGetValue(theirItem, out uint fairReceivedAmount) ? fairReceivedAmount + 1 : 1;
+
+										// Convert list to HashSet<Steam.Asset>
+										HashSet<Steam.Asset> fairItemsToGive = Trading.GetTradableItemsFromInventory(ourInventory, fairClassIDsToGive);
+										HashSet<Steam.Asset> fairItemsToReceive = Trading.GetTradableItemsFromInventory(theirInventory, fairClassIDsToReceive);
+
+										// Filter inventory for the sets we're looking for for IsTradeNeutralOrBetter
+										HashSet<Steam.Asset> fairFiltered = new HashSet<Steam.Asset>(theirInventory);
```

Comment:
That won't do, in fact, you didn't do anything productive over previous code, since you already initialized `fairFiltered` each time.

You need to make a copy of each item in inventory, so `IsTradeNeutralOrBetter()` can safely modify their amounts without affecting those in `theirInventory`. Returning to previous code and using `MemberwiseClone()` should be enough to achieve that.

Reply:
Isn't the constructor already giving me a copy with `new HashSet<Steam.Asset>(theirInventory)`? At least that's what I thought it did…

Reply:
Yes, it does, and you have two different sets pointing out to **the same items**, so when you modify amount of some item in set 2 then by definition you also modify the same item in set 1.

Reply:
Ah, great, it just copies the container, not the content.
Am I right I also have to do this for the failsafe in L594 then?

I think `fairFiltered.Add(item.MemberwiseClone());` is what you think of but it doesn't compile, telling me `error CS1540`.
~~~
Cannot access protected member 'object.MemberwiseClone()' via a qualifier of type 'Steam.Asset'; the qualifier must be of type 'Statistics' (or derived from it)`
~~~
I think that tells me that `MemberwiseClone()` only works within the same class?

Reply:
Just implement something like this in `Steam.Asset` and use it:

```
internal Asset ShallowCopy() {
    return (Asset) MemberwiseClone();
}
```

Reply:
And yes you'd need it for your failsafe as well, but you can remove it entirely as it's pointless, verifying A with A doesn't have any meaning, my failsafe was verifying my matching logic B with C verification whether the output was correct.